Standartize metadata for all objects on Adventure Map

ihhub opened the pull request:
- remove cryptic quantity1 and quantity2 values everywhere in the code and replace them with human-readable **metadata** array
- avoid hardcoded values for many object conditions like Artifact, Daemon Cave or Shipwreck
- fix AI pathfinding when Spell Scroll was considered as an artifact which requires to defeat an army
- fix Shipwreck Survivor bonuses. It always contained artifact of level 1
- add many checks in the code to prevent calling functions for unsupported objects
- add specialized functions to get information about objects. We have to avoid generic functions

relates to #6845 

**ATTENTION**
This pull request changes almost every object interaction on Adventure Map in relation to conditions and resources. Also I added a conversion code from old save format to a new one. As a result, it is important to test all objects for new maps as well as objects on maps loaded from old saves.

The best strategy for testing by developers is to enable debug mode and see metadata values on objects. Now it is easily readable.

ihhub commented:
Hi @oleg-derevenetz , @idshibanov and @Districh-ru , could you please review this pull request very carefully as there is a chance that I messed up somewhere. Once you are done with the review I will ask more people to review from the perspective of users.

ihhub commented:
So far I've tested most of stuff and it works fine for me. However, I don't have enough old save files to test everything.

Districh-ru reviewed (changes requested):
Hi, @ihhub.
I did a short test and got some bugs. Could you check it when you have time.
An assertion occurs when I try to open View World on map GRWAR2.MX2 (Great War 2). It happens in `Funds getDailyIncomeObjectResources( const Tiles & tile )` (file "maps_tiles_helper.cpp"):
`tile.GetObject( false )` returns **OBJ_RESOURCE (155)** for the tile #18415 and the tile actually has resource:
![изображение](https://user-images.githubusercontent.com/113276641/235483198-7dbfdd3c-5e14-4eff-8a93-1de2f806112b.png)
<details><summary>Tile info:</summary>
******* Tile info *******
Tile index      : 18415, point: (127, 127)
UID             : 8764
MP2 object type : 155 (Ресурс)
ICN object type : 46 (OBJNRSRC.ICN)
image index     : 5 (animated: 0)
layer type      : 0 - Object layer
region          : 8
ground          : Пустошь (isRoad: 0)
shadow          : false
passable from   : center, top, top-right, right, bottom-right, bottom, bottom-left, left
metadata value 1: 4
metadata value 2: 9
metadata value 3: 0
--------- Level 1 --------
UID             : 2559
ICN object type : 37 (OBJNTWSH.ICN)
image index     : 128
layer type      : 2 (2) - Shadow layer
is shadow       : yes
--------- Level 2 --------
UID             : 3025
ICN object type : 31 (MTNCRCK.ICN)
image index     : 25
layer type      : 0 (0) - Object layer
is shadow       : no
--- Extra information ---
*************************
</details>

This assertion occurs also on other started/loaded maps.

The same bug occurs when I try to load this save: [221_ghost_legion.zip](https://github.com/ihhub/fheroes2/files/11366462/221_ghost_legion.zip)

Districh-ru commented:
I checked View World fix and now it works correctly and the attached save file loads correctly.
I found another bug:
When I try to load [vh2_bad_mine.zip](https://github.com/ihhub/fheroes2/files/11366481/vh2_bad_mine.zip) I also get an assertion in `Maps::Tiles::quantityIntoMetadata()`.

ihhub commented:
Hi @Districh-ru , thank you very much for catching these bugs! I addressed all of them.

Districh-ru reviewed (changes requested):
Loading my save files I got assertions. could you check them also when you have time.
**1.**  [vh2_bad_mine.zip](https://github.com/ihhub/fheroes2/files/11366481/vh2_bad_mine.zip) I also get an assertion in `Maps::Tiles::quantityIntoMetadata()` (map_tiles.cpp)  - > `MP2::OBJ_SAWMILL`.
**2.** [A_Battle_of_four_0001.zip](https://github.com/ihhub/fheroes2/files/11366557/A_Battle_of_four_0001.zip) - in `quantityIntoMetadata()` - > `MP2::OBJ_DERELICT_SHIP`
**3.** [Biznaf_1_0065.zip](https://github.com/ihhub/fheroes2/files/11366562/Biznaf_1_0065.zip)  - in `quantityIntoMetadata()` - > `MP2::OBJ_SAWMILL`

Districh-ru commented:
After https://github.com/ihhub/fheroes2/pull/7084/commits/8482cdcbbee6f6111ec8279ed9bb02c0c09822cb I get only this bug:
[Biznaf_1_0065.zip](https://github.com/ihhub/fheroes2/files/11366562/Biznaf_1_0065.zip) and [3_Continents_0005.zip](https://github.com/ihhub/fheroes2/files/11366603/3_Continents_0005.zip) in "profit.cpp" `ProfitConditions::FromMine()`
All the rest of my save file load without an assertion. :)

ihhub commented:
Hi @Districh-ru , I fixed the last bug as well. Please check again.

Districh-ru commented:
> Hi @Districh-ru , I fixed the last bug as well. Please check again.

Well done!
Now all my saves load without assertions.

oleg-derevenetz commented on src/fheroes2/maps/maps_tiles.h:
```diff
@@ -493,12 +449,7 @@ namespace Maps
-
-        // Additional metadata is not set from map's information but during runtime like spells or monster joining conditions.
-        int32_t additionalMetadata = 0;
+        std::array<uint32_t, 3> _metadata{ 0 };
```
Why `std:array` and not `std::tuple`? As far as I can see, there are no such situations when index is not known at compile time. Am I wrong?

ihhub replied:
Hi @oleg-derevenetz , I honestly did not think about tuple :) I can change it to tuple.

oleg-derevenetz replied:
It's up to you :) There is `std::get<N>( T )` function that extracts N'th element from T, N is verified at compile time. It works also with arrays, but with arrays you can forget and write `T[N]`, while with tuple you cannot.

oleg-derevenetz reviewed (approved):
Hi @ihhub I checked savefiles that I have (mostly those from bug reports) and tried some new games with big maps, everything seems fine so far, but this of course requires more extensive testing. As for the code itself, I left one (optional) question, could you please take a look when possible?

Districh-ru commented on src/fheroes2/maps/maps_tiles_helper.cpp:
```diff
@@ -186,233 +381,192 @@ namespace Maps
 
-        assert( artifactId >= std::numeric_limits<Quantity1Type>::min() && artifactId <= std::numeric_limits<Quantity1Type>::max() );
+        switch ( static_cast<ArtifactCaptureCondition>( tile.metadata()[2] ) ) {
+        case ArtifactCaptureCondition::HAVE_WISDOM_SKILL:
+            return { Skill::Secondary::LEADERSHIP, Skill::Level::BASIC };
+        case ArtifactCaptureCondition::HAVE_LEADERSHIP_SKILL:
+            return { Skill::Secondary::WISDOM, Skill::Level::BASIC };
```
Maybe there is a misprint: the return values are swapped

ihhub replied:
Yes, it was a bug. It is fixed now.

Districh-ru commented on src/fheroes2/maps/maps_tiles_helper.cpp:
```diff
@@ -524,44 +671,78 @@ namespace Maps
+            return;
+        default:
+            // Why are you calling this function for an unsupported object type?
+            assert( 0 );
```
After defeating a monster in a New game I get this assertion: **tile.GetObject( false )** returns **OBJ_NONE (0)**

ihhub replied:
It is fixed now. We were deleting monster object from the map before setting metadata. This was a bug from the old code.

Districh-ru commented on src/fheroes2/maps/maps_tiles_helper.cpp:
```diff
@@ -186,233 +381,192 @@ namespace Maps
 
         case MP2::OBJ_FLOTSAM:
-            return Funds( Resource::GOLD, getGoldAmountFromTile( tile ) ) + Funds( Resource::WOOD, tile.GetQuantity1() );
+            return Funds{ Resource::WOOD, tile.metadata()[0] } + Funds{ Resource::GOLD, tile.metadata()[1] };
```
Should we multiply **tile.metadata()[1]** by 100 here?
As I understand we store gold in hundreds. It would be good to have a comment about maybe in Funds class.
Or could we do multiply the gold amount by 100 when we are putting it to `tile.metadata()[1]` and do not use `* 100` in every place we want to get gold amount from metadata.

ihhub replied:
Nope. We don't store Gold in hundreds anymore. Old format did but for simplification of debugging and development I remove this logic so Flotsam actually contains correct gold.

Districh-ru replied:
The previous case for MP2::OBJ_CAMPFIRE has `Funds{ Resource::GOLD, tile.metadata()[1] * 100' so I thought that we missed it here. :)

ihhub replied:
Campfire is the only object which has a connection between resource and gold. I checked the original game and if resource is N then the given gold is always N * 100.

Districh-ru replied:
So maybe It'll be good to leave this comment in code to have no questions in future. :)

ihhub replied:
It is done.

Districh-ru reviewed (changes requested):
Hi, @ihhub.
I have some bug reports and suggestions. Could you please take a look when you have time.

**1.** Picking up a any resource results in displaying that I picked **0** of the resource. But it is actually picked up as the amount of kingdom resource is increased:
![изображение](https://user-images.githubusercontent.com/113276641/235695802-aed3693b-3b4b-4dc2-88c0-0fdb3aaca6e8.png)

**2.** I cannot pick-up a bottle: [221_bottle_bug.zip](https://github.com/ihhub/fheroes2/files/11374566/221_bottle_bug.zip)
![изображение](https://user-images.githubusercontent.com/113276641/235702144-31ccfb8e-68ea-4b99-b934-33b4b8fd9710.png)

ihhub commented:
Hi @Districh-ru , I addressed all your comments except the bottle one. I don't know how it happens but it seems that the bottle actually doesn't exist there as it is showing OBJ_NONE. Did you start a new game for this?

Districh-ru commented:
> Hi @Districh-ru , I addressed all your comments except the bottle one. I don't know how it happens but it seems that the bottle actually doesn't exist there as it is showing OBJ_NONE. Did you start a new game for this?

Yes, I started a new game. In master built this bottle can be picked up. This is my test map: [221.zip](https://github.com/ihhub/fheroes2/files/11374857/221.zip)

ihhub commented:
Hi @Districh-ru , I fixed the bug and also some objects were missing on Adventure Map. Please take a look.

ihhub commented:
Hi @LeHerosInconnu, could you please help to test changes in this pull request? We need your comprehensive testing skills for this pull request.

LeHerosInconnu commented:
Hello @ihhub,

> Hi @LeHerosInconnu, could you please help to test changes in this pull request? We need your comprehensive testing skills for this pull request.

Are there any other changes planned before I start the test?

ihhub commented:
> Hello @ihhub,
> 
> > Hi @LeHerosInconnu, could you please help to test changes in this pull request? We need your comprehensive testing skills for this pull request.
> 
> Are there any other changes planned before I start the test?

Nope. I've recently fixed some issues found by @Branikolog and reported through Discord. As of now only bug fixing. You can download the newer version once the compilation is done.

oleg-derevenetz commented:
Hi @ihhub is this fine that when Tree of Knowledge has no resource requirement, there is "unknown resource" warning printed in the `Funds` constructor? It doesn't interfere with the game, but it looks weird and litters the log with messages about what is not really a problem.

ihhub commented:
> Tree of Knowledge

Hi @oleg-derevenetz , actually it is wrong. This message is added to avoid bad written code. I will have a look at it. Thank you for spotting it!

ihhub commented:
@oleg-derevenetz , fixed!

ihhub commented:
> What do you think about the following fixing method instead: allow to pass the `Resource::UNKNOWN` value to the `Funds` constructor (add the special `switch` case with just `break` and maybe also a check that `count` is zero as well and log a warning if it is not, because there cannot be non-zero amount of unknown resource)? Or this is undesirable?

I think we should prevent even passing unknown types into the constructor. This might help to find some future issues with wrong code or wrong data read from files. Of course no harm by doing this but in my view we should log such undesired behavior.

oleg-derevenetz commented:
Hi @ihhub there is the same issue in the `Maps::getFundsFromTile()`: non-pickable objects may contain no funds.

ihhub commented:
> Hi @ihhub there is the same issue in the `Maps::getFundsFromTile()`: non-pickable objects may contain no funds.

... I shall add the case with zero unknown resources :(

LeHerosInconnu commented:
Hello @ihhub,

I did a test on a test scenario and did not encounter any problems.
I will do another test playing a standard scenario.

ihhub commented:
Thank you @LeHerosInconnu !

Branikolog reviewed (approved):
Haven't found more bugs here. Great job, fheroes2 team! :D

Districh-ru reviewed (approved):
I played fully one new standard game and for a couple of week a campaign scenario and a standard game from an old save file. I did not notice any bugs.
Well done!

LeHerosInconnu commented:
Hello @ihhub,

> Hello @ihhub,
> 
> I did a test on a test scenario and did not encounter any problems.
> I will do another test playing a standard scenario.

I finished a standard scenario started with the PR version and continued with the merged version without any problems.
